system/lib/libcxx/src/new.cpp

```diff
@@ -90,10 +90,34 @@ operator new(std::size_t size) _THROW_BAD_ALLOC
     return p;
 }
 
+#if defined(__EMSCRIPTEN__) && defined(_LIBCPP_NO_EXCEPTIONS)
+void* _new_nothrow(size_t size) noexcept
+{
+    /// We cannot call ::operator new(size) here because it would abort
+    /// when malloc returns 0 and exceptions are disabled.
+    /// Expected behaviour of std::nothrow is not return 0 in that case.
+    void* p = nullptr;
+    if (size == 0)
+        size = 1;
+    while ((p = ::malloc(size)) == nullptr)
+    {
+        std::new_handler nh = std::get_new_handler();
+        if (nh)
+            nh();
+        else
+            break;
+    }
+    return p;
+}
```

Comment:
Sorry, I'm a little confused as to how to review this. What code does this replace? That is, I'm trying to find code to compare this against to see what it changes versus the "normal" code path.

Reply:
This is basically the same code as `void* operator new(std::size_t size) _THROW_BAD_ALLOC` above, just without the `#ifdef`s and without the Emscripten specific `abort`. The only code path that should be affected is when exceptions are disabled and the user explicitly uses e.g. `new (std::nothrow) char[123]` instead of `new char[123]`.

Reply:
I see, thanks, sorry for the late reply.

Comparing to that, I have another questions: Why is the emscripten-specific abort removed?

Reply:
I guess I'm confused as to what the connection is between throwing and not throwing and returning 0 and aborting.

Reply:
I think the idea is that if you explicitly call `new (std::nothrow)` then you are implying that it should return NULL and failure and not throw/trap/etc.

We don't have a separate settings for `ABORTING_NEW` like we do for `ABORTING_MALLOC`.  We historically would just always abort if allocation fails in `new`.   But I think `abort by default but not if std::nothrow is used` seems reasonable.   Our use of `abort` was essentially to replace the `throw` so it makes sense that the `nothrow` variant also wouldn't abort.

Reply:
I see. That seems reasonable.

This is a change though, so I think it should go in the changelog, and also perhaps in settings.js as a note on `ABORTING_MALLOC` (that new is different from malloc, and why).

Reply:
I updated the changelog and added a note to the `ABORTING_MALLOC` documentation. I'm really sorry that this took so long, but I forgot this after vacations. I also noticed that I used my work account for the commits. Since this PR is a bit outdated now, should I just rebase it onto the current main?

Reply:
Yes, looks like this has some conflicts that need to be resolved before landing. Merging is fine (we squash when we land).